mysql/datadog_checks/mysql/execution_plans.py

```diff
@@ -307,32 +312,106 @@ def _collect_execution_plans(self, db, tags, options, min_collection_interval):
+            1046,  # no permission on statement
+            1049,  # unknown database
+            1305,  # procedure does not exist
+            1370,  # no execute on procedure
+        })
 
         try:
-            start_time = time.time()
-            if schema is not None:
-                cursor.execute('USE `{}`'.format(schema))
-            cursor.execute('EXPLAIN FORMAT=json {statement}'.format(statement=statement))
-            statsd.timing("dd.mysql.run_explain.time", (time.time() - start_time) * 1000, tags=instance_tags)
-        except (pymysql.err.InternalError, pymysql.err.ProgrammingError) as e:
+            # Switch to the right schema; this is necessary when the statement uses non-fully qualified tables
+            # e.g. `select * from mytable` instead of `select * from myschema.mytable`
+            self._use_schema(cursor, schema)
+        except exceptions as e:
             if len(e.args) != 2:
                 raise
-            if e.args[0] in (1046,):
-                self.log.warning('Failed to collect EXPLAIN due to a permissions error: %s, Statement: %s', e.args,
-                                 statement)
-                return None
-            elif e.args[0] == 1064:
-                self.log.warning('Programming error when collecting EXPLAIN: %s, Statement: %s', e.args, statement)
-                return None
-            else:
-                raise
+            if e.args[0] in non_retryable_errors:
+                self._expiring_cache.set(strategy_cache_key, explain_strategy_none, strategy_cache_ttl)
+            self.log.debug('Cannot collect execution plan because %s schema could not be accessed: %s, statement: %s', schema, e.args, statement)
+            return None
+
+        # Use a cached strategy for the schema, if any, or try each strategy to collect plans
+        strategies = list(fns_by_strategy.keys())
+        cached = self._expiring_cache.get(strategy_cache_key)
+        if cached is not None:
+            strategies.remove(cached)
+            strategies.insert(0, cached)
+
+        for strategy in strategies:
+            fn = fns_by_strategy[strategy]
+
+            try:
+                plan = fn(cursor, statement)
+            except exceptions as e:
+                if len(e.args) != 2:
+                    raise
+                if e.args[0] in non_retryable_errors:
+                    self._expiring_cache.set(strategy_cache_key, explain_strategy_none, strategy_cache_ttl)
+                continue
+
+            if plan:
+                self._expiring_cache.set(strategy_cache_key, strategy, strategy_cache_ttl)
+                self.log.debug('Successfully collected plan using strategy: %s', strategy)
+                break
 
+        if not plan:
+            self.log.debug('Cannot collect execution plan because all strategies failed: %s', statement)
+
+        return plan
+
+    def _use_schema(self, cursor, schema):
+        if schema is not None:
```

Comment:
Probably shouldn't silently proceed here if the schema is `None`. Is this ever expected?

Reply:
Yes, this check was added because there were valid cases. You can connect to the database but have no schema associated with your session. As long as you only make fully-qualified queries, then all of those queries have no schema, but are valid queries.

Reply:
Makes sense. Maybe useful to add a comment explaining that